packages/block-editor/src/utils/dom.js

```diff
@@ -149,6 +149,14 @@ export function getVisibleElementBounds( element ) {
 		for ( const child of currentElement.children ) {
 			if ( isElementVisible( child ) ) {
 				const childBounds = child.getBoundingClientRect();
+				// If the child is larger than the parent, adjust the x and y values to account for any scrolling.
+				if ( childBounds.width > bounds.width ) {
+					childBounds.x = bounds.x;
+				}
+				if ( childBounds.height > bounds.height ) {
+					childBounds.y = bounds.y;
+					childBounds.height = bounds.height;
+				}
```

Comment:
Unfortunately I think the logic is imperfect — while it works for me when the links within a sub menu are the same height as the submenu item itself, if we add a link within the submenu that is taller (i.e. a link to a page that wraps to a second line) then the toolbar will overlap the submenu again:

![image](https://github.com/user-attachments/assets/0e35a998-e68a-40df-8323-6b04cc078706)

If I'm following the logic correctly, in this case, the individual item within the submenu is taller than `bounds.height` so we use `bounds.height` instead, so the bounds doesn't grow to include this item.

Apologies that this isn't a very useful comment — unfortunately I can't think of a neat way to determine if the child item is taller because it's in a popover (and so we want to include it) versus it being because of scroll (in which case we want to exclude it) 🤔

Reply:
Having trouble replicating it, but I see what you mean based on the logic you've explained

https://github.com/user-attachments/assets/c496cb68-6bc8-4e6f-83bd-303ae6eef73f

🤔 Do we only care about the parentRect?

Maybe we should rather be checking against that:

<details>

<summary>Example diff</summary>

```diff
diff --git a/packages/block-editor/src/utils/dom.js b/packages/block-editor/src/utils/dom.js
index 19dd24ce382..1983d37f723 100644
--- a/packages/block-editor/src/utils/dom.js
+++ b/packages/block-editor/src/utils/dom.js
@@ -140,7 +140,8 @@ export function getVisibleElementBounds( element ) {
 		return new window.DOMRectReadOnly();
 	}
 
-	let bounds = element.getBoundingClientRect();
+	const parentRect = element.getBoundingClientRect();
+	let bounds = parentRect;
 
 	const stack = [ element ];
 	let currentElement;
@@ -150,12 +151,12 @@ export function getVisibleElementBounds( element ) {
 			if ( isElementVisible( child ) ) {
 				const childBounds = child.getBoundingClientRect();
 				// If the child is larger than the parent, adjust the x and y values to account for any scrolling.
-				if ( childBounds.width > bounds.width ) {
-					childBounds.x = bounds.x;
+				if ( childBounds.width > parentRect.width ) {
+					childBounds.x = parentRect.x;
 				}
-				if ( childBounds.height > bounds.height ) {
-					childBounds.y = bounds.y;
-					childBounds.height = bounds.height;
+				if ( childBounds.height > parentRect.height ) {
+					childBounds.y = parentRect.y;
+					childBounds.height = parentRect.height;
 				}
 				bounds = rectUnion( bounds, childBounds );
 				stack.push( child );
```

</details>

Reply:
> Having trouble replicating it, but I see what you mean based on the logic you've explained

Now that you mention it, I can only replicate it when the tall menu item is the last in the list. If it isn't, then when the logic gets to the last child, the union will work against that item so the bug isn't apparent. So perhaps this is a bit more edge-casey than I thought! Ideally it'd be good to figure out, though 🤔

> 🤔 Do we only care about the parentRect?
> Maybe we should rather be checking against that:

Comparing explicitly against the `parentRect` now more closely matches the intent (and the comment), but I can still replicate the issue, so I'm not sure it changes much, unfortunately.

Reply:
I see it now - is it right that it only happens `onHover`?

Selecting the element, I cannot replicate.

Seems as if the position, onHover, is the previous position.

https://github.com/user-attachments/assets/3ec2fa6c-d566-4385-82b4-b7d9a322ae3b

Interesting!

Reply:
Or rather, I'm seeing it only when selecting a menu item with no submenu and the hovering over a menu item with a submenu.

Reply:
I managed to get something working by caching all the x/y values and using the max:


https://github.com/user-attachments/assets/2be9d5b3-2ebe-4926-8eb9-6d48400db2b5

It's pretty experimental, but I'll commit for consideration 😄 

https://github.com/user-attachments/assets/8fb35106-2eff-4fa6-b891-e80d71c07655

Reply:
Thanks for digging in further here! That's definitely getting it closer for me. Now it looks like the `y` position of the toolbar is directly beneath the last item, but not the padding of the sub menu. Much better but still not _quite_ right:

<img width="675" alt="image" src="https://github.com/user-attachments/assets/0b76dd86-5cd9-491f-9dc2-0aa62ac977ae">

But it might just be close enough?

> Or rather, I'm seeing it only when selecting a menu item with no submenu and the hovering over a menu item with a submenu.

Ah yep, that seems to be the case for me. That makes it feel even more edge-casey, so I feel like we could be _nearly_ close to something workable.